### PR TITLE
Add conferencing support

### PR DESCRIPTION
**Pull request by Kegsay**
This PR adds conferencing support to vector for HSes running https://github.com/matrix-org/matrix-appservice-verto (which matrix.org is)

Conferencing from the end-user's perspective:
 - Group chats may now have an "Ongoing conference call" notification on them.
 - Placing a voice or video call to the group chat room will no longer produce an error but will connect the client to a conference for that room.
 - Even if there is no "ongoing" notification, you can start a conference by just doing a voice/video call to the group chat room. This will invite a special conference user (part of the Verto AS) to mediate the conference call and provide ACLs on the conference call (so when you leave the group chat room, you get booted from the conference call). Semantics are the same as 1:1 calling (there is an end call button, etc).

Conferencing from a tech perspective:
 - The call is a faux 1:1 WebRTC call to the "special conference user" which has a programatically-determinable user ID of `@fs_$BASE-64-ROOM-ID:matrix.org`. The call is done in a 1:1 room with the conference user which is hidden from the recents list. In the future, we may want to expose this to provide floor control (muting/etc)
 - When the call is placed to this user, it will attempt to join the group chat room it is providing conferencing support to (it knows the room ID because its' user ID contains it!). It must be able to join the room in order for conferencing to work (this is why Vector by default will invite the user even if it is a public room).
 - When there are no more users on the conference call, the "special conference user" leaves the room. The state of the "special conference user" determines whether there is an "ongoing conference call" or not.

Additional notes:
 - This PR also fixes https://github.com/vector-im/vector-web/issues/31 - but the CSS for overlaying video needs to be done.
 - This PR also properly enforces single-call semantics (previously it was erroneously possible to place multiple calls(!))
 - There are a number of new configuration constants in SHOUTY_CAPS which determine the `user_id` format for the "special conference users" - we need to split this out into a config file for use at compile time ideally.

It has been manually tested for inbound and outbound calls, with changing rooms for each.

### CODE REVIEW

File: src/CallHandler.js [hunk 1]
```diff
@@ -161,40 +167,52 @@ dis.register(function(payload) {
...
+                }
+            }
+
             var members = room.getJoinedMembers();
-            if (members.length !== 2) {
-                var text = members.length === 1 ? "yourself." : "more than 2 people.";
+            if (members.length <= 1) {
                 Modal.createDialog(ErrorDialog, {
-                    description: "You cannot place a call with " + text
+                    description: "You cannot place a call with yourself."
```
Comment (ara4n) [inline src/CallHandler.js]: Why bother restricting this?  It provides a simple test service out of the box - "to test VoIP, just start a room and start a VC in it and check you see your reflection!  Then invite more people in and have fun!"

Comment (ara4n) [inline src/CallHandler.js]: ...oh, because I guess migrating from 1 user to 1:1 (2 users) to many users would be a PITA.  Boo.  Guess we'll need a proper test service instead, or manage that dance more nicely.

**Comment by ara4n** [comment]
lgtm on first inspection, but need to fix the CSS before we can land it. doing so now.

File: skins/base/css/templates/Login.css [hunk 2]
```diff
@@ -75,6 +75,22 @@ limitations under the License.
...
+    text-align: center;
+    width: 100%;
+    font-size: 14px;
+    opacity: 0.8;
+}
+
+.mx_Login_links a:link {
+   color: #4a4a4a;
+}
+
```
Comment (dbkr) [inline skins/base/css/templates/Login.css]: Is this meant to be in a PR for conferencing support?

File: skins/base/views/templates/Login.js [hunk 3]
```diff
@@ -165,6 +165,13 @@ module.exports = React.createClass({
                         {this.state.errorText}
                 </div>
                 <a className="mx_Login_create" onClick={this.showRegister} href="#">Create a new account</a>
+                <br/>
+                <div className="mx_Login_links">
+                    <a href="https://medium.com/@Vector">blog</a>&nbsp;&nbsp;&middot;&nbsp;&nbsp;
+                    <a href="https://twitter.com/@VectorCo">twitter</a>&nbsp;&nbsp;&middot;&nbsp;&nbsp;
+                    <a href="https://github.com/vector-im/vector-web">github</a>&nbsp;&nbsp;&middot;&nbsp;&nbsp;
+                    <a href="https://matrix.org">powered by Matrix</a>
+                </div>
```
Comment (dbkr) [inline skins/base/views/templates/Login.js]: Oh, this explains the style above. This should not be part of this PR.

**Comment by dbkr** [comment]
I don't think this should go into master because:

1) There's no matching specification in Matrix for it so it would be a custom extension
2) In its current state it's limited to only using the conferencing AS on matrix.org

Suggestions to move forward:

 * If this is a functionality demo, a better place for it would be on a branch or a fork.
 * Otherwise, start proposing a standard for multi party VoIP

**Comment by ara4n** [comment]
For now this is a vector specific solution, and it works on all HSes via the matrix.org AS. Why wouldn't we put it on vector master? Vector isn't a reference impl...

It doesn't need any spec, as it's not yet doing anything custom with Matrix - no custom events or VoIP. It's just one client (vector) talking to another (freeswitch via verto).

If we want to formalise this way of doing conferencing then we could add it to the underlying matrix sdks (eg matrix-react-sdk).

**Comment by Kegsay** [comment]
From my point of view:
>  the only thing which is unofficial is the syntax for inviting the AS user - the act of placing a call to a user remains the same. For example, I can and have done conf calls on android which has no special code to handle this, simply by doing a 1:1 call to the @fs_ user [...] the ux is just bad on other clients

I also have to agree that this is Vector, not a reference implementation. I don't think spec enforcements for a non-official client is proportional or relevant, but then again that isn't for me to decide. When some conclusion has been reached, I'll apply any actionable things to this PR.

File: skins/base/views/templates/Login.js (hunk 3 above)
Comment (ara4n) [inline skins/base/views/templates/Login.js]: as I said on Matrix internal at the time - this was a bodged rushed commit on the wrong branch. if you really want me to back it out and put it in its own PR I will, but it seems like bureaucracy for the sake of it...

**Comment by dbkr** [comment]
Accepting since this now has some doc explaining the context, and on the basis that it's a vector-specific impl of one possible way of doing conf in Matrix.
